### PR TITLE
PRO-3409 named anchors

### PR DESCRIPTION
**Pull request by boutell**
## Summary

Implemented named anchor editing. Named anchors are implemented as spans with an `id` attribute, because in our tiptap-based editor everything with editable attributes has to be eventually wrapped as a tag. Existing `<a name>` tags are migrated. If they have no children, the new span is automatically modified to wrap the next text node, so that there is something for our editor to recognize. Otherwise the span gets discarded by tiptap.

In an extreme edge case where there is no next text node in traversal of the whole document an anchor emoji is inserted but this shouldn't happen.

## What are the specific steps to test this change?

Include `anchor` on the toolbar. Mark some text that is a few screens down the page as a named anchor, then make a link to it somewhere else using `#anchorname`. That link should work after you publish and load the page in an incognito window. Or in preview mode, probably.

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [X] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [X] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [X] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

Screenshot 2022-12-15 at 5 44 36 PM
Screenshot 2022-12-15 at 5 44 40 PM

### CODE REVIEW

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 1]
```diff
@@ -445,13 +498,13 @@ module.exports = {
         const output = await _super(req, input, rteOptions);
         const finalOptions = self.optionsToSanitizeHtml(rteOptions);
 
-        output.content = sanitizeHtml(input.content, finalOptions);
+        output.content = self.sanitizeHtml(input.content, finalOptions);
 
         const anchors = output.content.match(/"#apostrophe-permalink-[^"?]*?\?/g);
-        output.permalinkIds = anchors.map(anchor => {
+        output.permalinkIds = (anchors && anchors.map(anchor => {
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/index.js:504]: Unrelated fix for the permalinks PR, which this will merge into.

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposRichTextWidgetEditor.vue [hunk 2]
```diff
@@ -298,6 +291,54 @@ export default {
...
+    // Legacy content may have `id` and `name` attributes on anchor tags
+    // but our tiptap anchor extension needs them on a separate `span`, so nest
+    // a span to migrate this content for each relevant anchor tag encountered
+    transformNamedAnchors(html) {
+      const el = document.createElement('div');
+      el.innerHTML = html;
+      const anchors = el.querySelectorAll('a[name]');
+      for (const anchor of anchors) {
+        const name = anchor.getAttribute('id') || anchor.getAttribute('name');
+        if (((typeof name) === 'string') && name.length) {
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposRichTextWidgetEditor.vue:303]: could early `continue`  to reduce indentation

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposRichTextWidgetEditor.vue [hunk 3]
```diff
@@ -298,6 +291,54 @@ export default {
...
+          const span = document.createElement('span');
+          span.setAttribute('id', name);
+          anchor.removeAttribute('id');
+          anchor.removeAttribute('name');
+          if (anchor.children.length) {
+            // Migrate children of the anchor to the span
+            while (anchor.firstElementChild) {
+              span.append(anchor.firstElementChild);
+            }
+            anchor.prepend(span);
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposRichTextWidgetEditor.vue:313]: could we [replace](https://developer.mozilla.org/en-US/docs/Web/API/Element/replaceWith) the old anchor instead of appending the new one in it?

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+      }"
+    >
+      <AposContextMenuDialog
+        menu-placement="bottom-start"
+      >
+        <div v-if="hasAnchorOnOpen" class="apos-anchor-control__remove">
+          <AposButton
+            type="quiet"
+            @click="removeAnchor"
+            label="apostrophe:removeAnchor"
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue:29]: missing translation

image

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+          :trigger-validation="triggerValidation"
+          v-model="docFields"
+          :utility-rail="false"
+          :modifiers="formModifiers"
+          :key="lastSelectionTime"
+          :generation="generation"
+          :following-values="followingValues()"
+          :conditional-fields="conditionalFields()"
+        />
+        <footer class="apos-link-control__footer">
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue:43]:
```suggestion
        <footer class="apos-anchor-control__footer">
```

File: modules/@apostrophecms/rich-text-widget/ui/apos/tiptap-extensions/Anchor.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+        unsetAnchor: () => ({ chain }) => {
+          return chain()
+            .unsetMark(this.name, { extendEmptyMarkRange: true })
+            .run();
+        }
+      };
+    }
+  });
+};
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/tiptap-extensions/Anchor.js:59]: new line

**Review by ETLaurent — CHANGES_REQUESTED** [review]
Tested on testbed, works well!

Could `AposTiptapLink.vue` and `AposTiptapAnchor.vue` share their common code in a mixin? 
Or is it urgent? (and could be done in another time)

**Review by ETLaurent — CHANGES_REQUESTED** [review]
just the missing translation and [this comment](https://github.com/apostrophecms/apostrophe/pull/3967/files#r1050761464), and it's good to me!

image

**Review by ETLaurent — COMMENTED** [review]
2 missing keys actually

image

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+      hasAnchorOnOpen: false,
+      triggerValidation: false,
+      docFields: {
+        data: {}
+      },
+      formModifiers: [ 'small', 'margin-micro' ],
+      originalSchema: [
+        {
+          name: 'anchor',
+          label: 'Anchor Name',
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue:93]: should be a i18n key, what do you think?

**Review by ETLaurent — CHANGES_REQUESTED** [review]

**Review by ETLaurent — APPROVED** [review]
